File: android/richtext/src/main/kotlin/com/abracode/richtext/find/RichTextFindController.kt
package com.abracode.richtext.find

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Rect
import com.abracode.richtext.model.RichTextDocument
import com.abracode.richtext.rendering.RichTextHighlightStyle
import com.abracode.richtext.rendering.RichTextHighlights
import com.abracode.richtext.rendering.RichTextRenderedText
import com.abracode.richtext.search.RichTextRange
import com.abracode.richtext.search.RichTextSearch
import com.abracode.richtext.search.RichTextSearchOptions

// Port of Sources/RichText/Find/RichTextFindController.swift - layer 3: the state behind a find bar over ONE
// document. It owns the query and options, runs the engine over the rendered text it is bound to, and tracks
// which match the reader is on. It knows nothing about composables: RichTextFindBar edits it, `RichText(find =
// controller)` binds a document to it and paints its highlights, and a host with its own search field can drive
// `query` directly and never show the bar. Snapshot state (mutableStateOf) rather than StateFlow, like the
// ChatView store, so composables observe it for free.
//
// It keeps ranges, not matches: a bar shows "3 of 12" and paints, never a snippet.
class RichTextFindController {

    private var queryState by mutableStateOf("")
    private var optionsState by mutableStateOf(RichTextSearchOptions.Default)

    /** The text to find. Setting it re-runs the search; "" clears (the engine matches nothing for it). */
    var query: String
        get() = queryState
        set(value) {
            if (queryState == value) return
            queryState = value
            recompute()
        }

    var options: RichTextSearchOptions
        get() = optionsState
        set(value) {
            if (optionsState == value) return
            optionsState = value
            recompute()
        }

    /** Whether the bar is shown. The controller only stores this; the host's layout reads it. */
    var isPresented by mutableStateOf(false)

    /**
     * Bumped by every `present(focus = true)`, so a repeat request while the bar is open still puts focus back in
     * the field. The bar records the request it honored in [focusHonored], so a bar re-created later does not
     * honor it twice.
     */
    var focusRequests by mutableStateOf(0)
        private set
    var focusHonored by mutableStateOf(0)
        private set

    /** Colors for the painted ranges. */
    var style by mutableStateOf(RichTextHighlightStyle.Default)

    var ranges: List<RichTextRange> by mutableStateOf(emptyList())
        private set

    /** Index into [ranges] of the match the reader is on; null when there are none. */
    var currentIndex: Int? by mutableStateOf(null)
        private set

    /**
     * The current match's bounds in WINDOW coordinates, reported by the RichText composable after layout (null
     * until it is laid out, or when there is no current match). A host scroller brings it into view from this.
     */
    var currentMatchBounds: Rect? by mutableStateOf(null)
        internal set

    /** The rendered text being searched (bound by `RichText(find = ...)`, or by hand for a headless host). */
    var text: String by mutableStateOf("")
        private set

    // --- Binding to a document. ---

    /** Search this rendered text from now on; the cursor is kept when its range survives, else back to the first. */
    fun bind(text: String) {
        if (text == this.text) return
        this.text = text
        recompute()
    }

    /** Bind to a document: the ranges the controller finds are the ranges the composable paints. */
    fun bind(document: RichTextDocument) = bind(RichTextRenderedText.layout(document).text)

    // --- Navigation. ---

    fun next() = step(1)

    fun previous() = step(-1)

    /** Jump to a specific match (a result-list tap). */
    fun select(index: Int) {
        if (index in ranges.indices) currentIndex = index
    }

    /** Hide the bar and clear the query, so nothing stays painted. */
    fun dismiss() {
        isPresented = false
        query = ""
    }

    /**
     * Show the bar. With [focus] (the reader's own gesture) its field takes focus, again if the bar is already
     * open; without it (a host driving [query] from its own search field) the bar appears and focus stays where
     * the reader is typing.
     */
    fun present(focus: Boolean = true) {
        isPresented = true
        if (focus) focusRequests += 1
    }

    /** The bar took the focus [focusRequests] asked for. */
    fun markFocusHonored() {
        focusHonored = focusRequests
    }

    // --- Derived. ---

    /** What `RichText(highlights = ...)` should paint: null when there is nothing to paint. */
    val highlights: RichTextHighlights?
        get() = if (ranges.isEmpty()) null else RichTextHighlights(ranges, currentIndex, style)

    /** "3 of 12", "No matches", or "" for an empty query. With `options.limit` reached, "3 of 500+". */
    val summary: String
        get() {
            if (queryState.isBlank()) return ""
            val index = currentIndex ?: return "No matches"
            if (ranges.isEmpty()) return "No matches"
            val capped = optionsState.limit?.let { ranges.size >= it } ?: false
            return "${index + 1} of ${ranges.size}${if (capped) "+" else ""}"
        }

    // --- Private. ---

    private fun step(delta: Int) {
        if (ranges.isEmpty()) return
        val count = ranges.size
        val base = currentIndex ?: if (delta > 0) -1 else 0
        currentIndex = ((base + delta) % count + count) % count
    }

    private fun recompute() {
        val previousRange = currentIndex?.let { ranges.getOrNull(it) }
        ranges = RichTextSearch.ranges(text, query, options)
        val kept = previousRange?.let { ranges.indexOf(it) } ?: -1
        currentIndex = when {
            ranges.isEmpty() -> null
            kept >= 0 -> kept
            else -> 0
        }
        if (currentIndex == null) currentMatchBounds = null
    }
}

File: Sources/RichText/Find/RichTextFindController.swift
// Sources/RichText/Find/RichTextFindController.swift
//
// Layer 3 of the find stack (see RichTextSearch.swift): the state behind a find bar over ONE document. It
// owns the query and options, runs the engine over the rendered text it is bound to, and tracks which match
// the user is on. It knows nothing about views: RichTextFindBar edits it, `RichText.find(_:)` paints its
// `highlights`, and a host with its own search field can drive `query` directly and never show the bar.
//
// A transcript of many documents does not use this - each message is its own RichText, and the transcript
// keeps one cross-message cursor of its own, handing each message just its ranges via `findHighlights`.
// ObservableObject (not @Observable): RichText's floor is macOS 13 / iOS 16.
//
// It keeps ranges, not RichTextMatch values: a bar shows "3 of 12" and paints, and never displays a
// snippet, so building one per hit on every keystroke would be waste. A result list wants
// `RichTextSearch.matches` instead.

import Foundation
import Combine

@MainActor
public final class RichTextFindController: ObservableObject {

    /// The text to find. Setting it re-runs the search; "" clears matches (the engine matches nothing
    /// for an empty query, so the document never lights up wholesale).
    @Published public var query: String = "" {
        didSet {
            if query != oldValue {
                recompute()
            }
        }
    }

    @Published public var options: RichTextSearchOptions = .default {
        didSet {
            if options != oldValue {
                recompute()
            }
        }
    }

    /// Whether the bar is shown. The controller only stores this; `richTextFindBar` reads it.
    @Published public var isPresented = false

    /// Bumped by `present(focus: true)` every time, so a repeat Cmd-F while the bar is already open
    /// still puts focus back in the field (a plain Bool would not change, and nothing would be
    /// published). The bar records the request it honored in `focusHonored`.
    @Published public private(set) var focusRequests = 0
    @Published public private(set) var focusHonored = 0

    /// Colors for the painted ranges.
    @Published public var style: RichTextHighlightStyle = .default

    @Published public private(set) var ranges: [NSRange] = []

    /// Index into `ranges` of the match the user is on; nil when there are none.
    @Published public private(set) var currentIndex: Int?

    /// The rendered text being searched (bound by `RichText.find(_:)`, or by hand for a headless host).
    public private(set) var text: String = ""

    public init() {}

    // MARK: - Binding to a document

    /// Search this rendered text from now on. The current match is kept when its range survives (the same
    /// text was re-bound), else the cursor goes back to the first match.
    public func bind(text: String) {
        guard text != self.text else {
            return
        }
        self.text = text
        recompute()
    }

    public func bind(attributed: NSAttributedString) {
        bind(text: attributed.string)
    }

    /// Bind to a document rendered with the same theme / engine as the view showing it, so the ranges the
    /// controller finds are the ranges the view paints.
    public func bind(document: RichTextDocument, theme: RichTextTheme = .default, engine: RichTextEngine = .textKit1) {
        bind(attributed: RichTextAttributedString.make(document, theme: theme, engine: engine))
    }

    // MARK: - Navigation

    public func next() {
        step(by: 1)
    }

    public func previous() {
        step(by: -1)
    }

    /// Jump to a specific match (a result-list tap).
    public func select(_ index: Int) {
        guard ranges.indices.contains(index) else {
            return
        }
        currentIndex = index
    }

    /// Hide the bar and clear the query, so nothing stays painted.
    public func dismiss() {
        isPresented = false
        query = ""
    }

    /// Show the bar. With `focus` (the reader's Cmd-F) its field takes focus, again if the bar is
    /// already open; without it (a host driving `query` from its own search field) the bar appears and
    /// focus stays where the reader is typing.
    public func present(focus: Bool = true) {
        isPresented = true
        if focus {
            focusRequests += 1
        }
    }

    /// The bar took the focus `focusRequests` asked for.
    public func markFocusHonored() {
        focusHonored = focusRequests
    }

    // MARK: - Derived

    /// What `RichText.findHighlights` should paint: nil when there is nothing to paint.
    public var highlights: RichTextHighlights? {
        guard !ranges.isEmpty else {
            return nil
        }
        return RichTextHighlights(ranges: ranges, current: currentIndex, style: style)
    }

    /// "3 of 12", "No matches", or "" for an empty query. With `options.limit` reached, "3 of 500+".
    public var summary: String {
        if query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return ""
        }
        guard let currentIndex, !ranges.isEmpty else {
            return "No matches"
        }
        let capped = options.limit.map { ranges.count >= $0 } ?? false
        return "\(currentIndex + 1) of \(ranges.count)\(capped ? "+" : "")"
    }

    // MARK: - Private

    private func step(by delta: Int) {
        guard !ranges.isEmpty else {
            return
        }
        let count = ranges.count
        let base = currentIndex ?? (delta > 0 ? -1 : 0)
        currentIndex = ((base + delta) % count + count) % count
    }

    private func recompute() {
        let previousRange = currentIndex.flatMap { ranges.indices.contains($0) ? ranges[$0] : nil }
        ranges = RichTextSearch.ranges(in: text, query: query, options: options)
        if ranges.isEmpty {
            currentIndex = nil
        } else if let previousRange, let kept = ranges.firstIndex(of: previousRange) {
            currentIndex = kept
        } else {
            currentIndex = 0
        }
    }
}
